In [81]:
import pandas as pd
import numpy as np
import sys
import re

import seaborn as sns
import matplotlib.pyplot as plt

In [82]:
sys.path.append('../')
import src.limpieza as lmp

In [83]:
import requests

In [84]:
pd.options.display.max_columns = None

In [85]:
url = 'https://www.ine.es/jaxiT3/files/t/es/csv_bdsc/2078.csv'
res1 = requests.get(url)


In [86]:
res1.status_code

200

In [87]:
from pathlib import Path

data_folder = Path(r"C:\Users\mituc\Ironhack\Curso\IronLabs\Proyecto-1\data")
file_to_open = data_folder / "2078.csv"
with open(file_to_open, 'wb') as f:
    f.write(res1.content)

In [88]:
df_guiris = pd.read_csv(r"C:\Users\mituc\Ironhack\Curso\IronLabs\Proyecto-1\data\2078.csv", sep=';', encoding='latin-1', index_col = 0)

In [89]:
df_guiris.shape

(86920, 4)

In [90]:
df_guiris.head(5)

Viajeros y pernoctaciones             Residencia  \
Puntos turísticos                                                    
01059 Vitoria-Gasteiz                   Viajero  Residentes en España   
01059 Vitoria-Gasteiz                   Viajero  Residentes en España   
01059 Vitoria-Gasteiz                   Viajero  Residentes en España   
01059 Vitoria-Gasteiz                   Viajero  Residentes en España   
01059 Vitoria-Gasteiz                   Viajero  Residentes en España   

                       Periodo   Total  
Puntos turísticos                   
01059 Vitoria-Gasteiz  2022M01  11.759  
01059 Vitoria-Gasteiz  2021M12  16.496  
01059 Vitoria-Gasteiz  2021M11  16.667  
01059 Vitoria-Gasteiz  2021M10  21.467  
01059 Vitoria-Gasteiz  2021M09  18.647

In [91]:
df_guiris.duplicated().sum()

12322

It seems that we have quite a lot of duplicates but, is it true?

In [92]:
df_guiris[df_guiris.duplicated() == True]

Viajeros y pernoctaciones                   Residencia  \
Puntos turísticos                                                          
02003 Albacete                          Viajero        Residentes en España   
02003 Albacete                          Viajero        Residentes en España   
02003 Albacete                          Viajero        Residentes en España   
02003 Albacete                          Viajero  Residentes en el Extranjero   
02003 Albacete                          Viajero  Residentes en el Extranjero   
...                                         ...                          ...   
50297 Zaragoza                   Pernoctaciones        Residentes en España   
50297 Zaragoza                   Pernoctaciones        Residentes en España   
50297 Zaragoza                   Pernoctaciones  Residentes en el Extranjero   
50297 Zaragoza                   Pernoctaciones  Residentes en el Extranjero   
50297 Zaragoza                   Pernoctaciones  Residentes en el Extranjero   

                       Periodo Total  
Puntos turísticos                 
02003 Albacete         2020M06     .  
02003 Albacete         2020M05     .  
02003 Albacete         2020M04     0  
02003 Albacete         2020M06     .  
02003 Albacete         2020M05     .  
...                        ...   ...  
50297 Zaragoza         2020M05     .  
50297 Zaragoza         2020M04     0  
50297 Zaragoza         2020M06     .  
50297 Zaragoza         2020M05     .  
50297 Zaragoza         2020M04     0  

[12322 rows x 4 columns]

In [93]:
df_guiris[df_guiris['Total'] == '0']

Viajeros y pernoctaciones                   Residencia  \
Puntos turísticos                                                          
01059 Vitoria-Gasteiz                   Viajero        Residentes en España   
01059 Vitoria-Gasteiz                   Viajero  Residentes en el Extranjero   
01059 Vitoria-Gasteiz            Pernoctaciones        Residentes en España   
01059 Vitoria-Gasteiz            Pernoctaciones  Residentes en el Extranjero   
02003 Albacete                          Viajero        Residentes en España   
...                                         ...                          ...   
49275 Zamora                     Pernoctaciones  Residentes en el Extranjero   
50297 Zaragoza                          Viajero        Residentes en España   
50297 Zaragoza                          Viajero  Residentes en el Extranjero   
50297 Zaragoza                   Pernoctaciones        Residentes en España   
50297 Zaragoza                   Pernoctaciones  Residentes en el Extranjero   

                       Periodo Total  
Puntos turísticos                 
01059 Vitoria-Gasteiz  2020M04     0  
01059 Vitoria-Gasteiz  2020M04     0  
01059 Vitoria-Gasteiz  2020M04     0  
01059 Vitoria-Gasteiz  2020M04     0  
02003 Albacete         2020M04     0  
...                        ...   ...  
49275 Zamora           2020M04     0  
50297 Zaragoza         2020M04     0  
50297 Zaragoza         2020M04     0  
50297 Zaragoza         2020M04     0  
50297 Zaragoza         2020M04     0  

[425 rows x 4 columns]

Although it seemed that there were duplicates, in the previous table we can see that this is not the case and if we reset the index they disappear

In [94]:
df_guiris.reset_index(inplace=True)

In [95]:
df_guiris.duplicated().sum()

0

In [96]:
df_guiris.dtypes

Puntos turísticos        object
Viajeros y pernoctaciones    object
Residencia                   object
Periodo                      object
Total                        object
dtype: object

We'll have to clean the column names

In [97]:
df_guiris.columns

Index(['Puntos turísticos', 'Viajeros y pernoctaciones', 'Residencia',
       'Periodo', 'Total'],
      dtype='object')

In [98]:
print(df_guiris['Viajeros y pernoctaciones'].unique())
print(df_guiris['Residencia'].unique())

['Viajero' 'Pernoctaciones']
['Residentes en España' 'Residentes en el Extranjero']


In [99]:
df_guiris.isnull().sum()

Puntos turísticos           0
Viajeros y pernoctaciones       0
Residencia                      0
Periodo                         0
Total                        5358
dtype: int64

In [100]:
df_guiris.shape

(86920, 5)

We rename the columns

In [101]:
df_guiris = df_guiris.rename(columns={'Puntos turísticos': 'Puntos_turisticos'})
columnas_limpias = {col : col.lower().replace(" ", "_") for col in df_guiris.columns}
df_guiris.rename(columns = columnas_limpias, inplace = True)

In [102]:
df_guiris.columns

Index(['puntos_turisticos', 'viajeros_y_pernoctaciones', 'residencia',
       'periodo', 'total'],
      dtype='object')

Now let's clean up the 'total' column. We will first remove the dots and then convert the values ​​to integers.

It is important to note that in this csv if there is dot '.' alone, it may mean that the data is protected by the statistical secrecy law (Article 12 of Law 7/2006) or that the data was not collected. Both cases will be treated as if they were nulls.

This cleaning operation has therefore a double function:
- be able to convert strings to int
- convert the '.' in NaN to handle them later

In [103]:
df_guiris['total'] = df_guiris['total'].str.replace('.', '')

C:\Users\mituc\AppData\Local\Temp/ipykernel_11116/2279084469.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  df_guiris['total'] = df_guiris['total'].str.replace('.', '')


In [104]:
df_guiris['total'] = pd.to_numeric(df_guiris['total'], errors='coerce').astype('Int64')

We can see that the nulls went up from 5358 to 13056.

In [105]:
df_guiris.isnull().sum()

puntos_turisticos                0
viajeros_y_pernoctaciones        0
residencia                       0
periodo                          0
total                        13056
dtype: int64

We know by the documentation of the csv that between april and june 2020 there is no data due to the covid-19 crisis. Let's take a look at that period.

In [106]:
df_guiris[(df_guiris['periodo'] == '2020M04') | (df_guiris['periodo'] == '2020M05') | (df_guiris['periodo'] == '2020M06')]

puntos_turisticos viajeros_y_pernoctaciones  \
19     01059 Vitoria-Gasteiz                   Viajero   
20     01059 Vitoria-Gasteiz                   Viajero   
21     01059 Vitoria-Gasteiz                   Viajero   
224    01059 Vitoria-Gasteiz                   Viajero   
225    01059 Vitoria-Gasteiz                   Viajero   
...                      ...                       ...   
86530         50297 Zaragoza            Pernoctaciones   
86531         50297 Zaragoza            Pernoctaciones   
86734         50297 Zaragoza            Pernoctaciones   
86735         50297 Zaragoza            Pernoctaciones   
86736         50297 Zaragoza            Pernoctaciones   

                        residencia  periodo  total  
19           Residentes en España  2020M06   <NA>  
20           Residentes en España  2020M05   <NA>  
21           Residentes en España  2020M04      0  
224    Residentes en el Extranjero  2020M06   <NA>  
225    Residentes en el Extranjero  2020M05   <NA>  
...                            ...      ...    ...  
86530        Residentes en España  2020M05   <NA>  
86531        Residentes en España  2020M04      0  
86734  Residentes en el Extranjero  2020M06   <NA>  
86735  Residentes en el Extranjero  2020M05   <NA>  
86736  Residentes en el Extranjero  2020M04      0  

[1272 rows x 5 columns]

As during the covid-19 crisis all the turistic activity was suspended, we will substitute the nulls in those months by 0, as the problem wasn't in the data collection, but due to the fact that there were no tourist at all.

In [107]:
covid = ['2020M04', '2020M05', '2020M06']

df_guiris.loc[(df_guiris['periodo'].isin(covid)), 'total'] = 0

And we check that the nulls went down from 13056 to 12208

In [108]:
df_guiris.isnull().sum()

puntos_turisticos                0
viajeros_y_pernoctaciones        0
residencia                       0
periodo                          0
total                        12208
dtype: int64

The nulls that we still have are due to two facts: there wasn't any data collection for place at those dates or the data is protected by statistic secretc law. Anyway, we cannot erase them directly; 
in the event that they are missing in specific months but we have data for the rest of the year, it may be convenient to replace them by the mean or the median.

In [109]:
null_rows = pd.isnull(df_guiris['total'])

In [110]:
df_guiris[null_rows].to_csv('nulls.csv')

In [111]:
df_guiris['periodo'] = df_guiris['periodo'].str.replace('M', '-')

In [112]:
df_guiris.to_pickle(r'C:\Users\mituc\Ironhack\Curso\IronLabs\Proyecto-1\data\guiris.pkl')
#df_guiris = pd.read_pickle(r'C:\Users\mituc\Ironhack\Curso\IronLabs\Proyecto-1\data\guiris.pkl')


In [113]:
df_guiris['periodo'] = pd.to_datetime(df_guiris['periodo'], format='%Y%m%')
df_guiris['periodo'] = df_guiris['periodo'].dt.date.apply(lambda x: x.strftime('%Y-%m'))

In [114]:
df_guiris.dtypes

puntos_turisticos            object
viajeros_y_pernoctaciones    object
residencia                   object
periodo                      object
total                         Int64
dtype: object

In [115]:
df_guiris.set_index('periodo', inplace=True)

We rename the values

In [116]:
df_guiris['residencia'].replace({'Residentes en España': "residentes_espana", 'Residentes en el Extranjero': 'residentes_en_el_extranjero'}, inplace=True)
df_guiris['viajeros_y_pernoctaciones'] = df_guiris['viajeros_y_pernoctaciones'].str.lower()

Aplicamos la función que limpia y organiza los datos por ciudades

In [117]:
lista = ['Barcelona']

In [118]:
lmp.pastor_de_guiris(df_guiris, lista)

In [119]:
df_guiris = pd.read_pickle(r'C:\Users\mituc\Ironhack\Curso\IronLabs\Proyecto-1\data\Barcelona_guiris_esquilados.pkl')

In [120]:
df_guiris.head(5)

viajeros_espana  viajeros_extranjero  pernoctaciones_espana  \
periodo                                                                
2022-01           107162               194312                 203683   
2021-12           144299               250133                 282398   
2021-11           120748               314677                 228675   
2021-10           130404               364349                 270494   
2021-09           102053               309572                 207329   

         pernoctaciones_extranjero  media_pernoc_esp  media_pernoc_ext  \
periodo                                                                  
2022-01                     541541               1.9               2.8   
2021-12                     674630               2.0               2.7   
2021-11                     871021               1.9               2.8   
2021-10                    1013514               2.1               2.8   
2021-09                     856562               2.0               2.8   

         total_viajeros  total_pernoctaciones  media_total_pernoc  
periodo                                                            
2022-01          301474                745224                 2.5  
2021-12          394432                957028                 2.4  
2021-11          435425               1099696                 2.5  
2021-10          494753               1284008                 2.6  
2021-09          411625               1063891                 2.6